### PR TITLE
[NEXUS-8473] customize shiro session-factory to ensure attributes map is threadsafe

### PR DESCRIPTION
**Pull request by jdillon**
https://issues.sonatype.org/browse/NEXUS-8473

### CODE REVIEW

**Comment by jdillon** [comment]
appears may be causing some issues, will look more.... fun!

**Comment by jdillon** [comment]
looks like it may be a simple change to SimpleSession to wrap the HashMap with Collections.synchronizedMap() but the current impl of SimpleSession looks like it requires us to replace the entire class to make this one change.

File: components/nexus-security/src/main/java/org/apache/shiro/nexus/NexusSessionFactory.java [hunk 1]
```diff
@@ -29,9 +33,49 @@
...
+     *
+     * This appears to only be called by {@link SimpleSession#getAttributesLazy()}.
+     */
+    @Override
+    public void setAttributes(final Map<Object, Object> attributes) {
+      // NOTE: usage seems to be non-null, but guard for sanity
+      if (attributes != null) {
+        super.setAttributes(Collections.synchronizedMap(attributes));
+      }
+      super.setAttributes(null);
```
Comment (mcculls) [inline components/nexus-security/src/main/java/org/apache/shiro/nexus/NexusSessionFactory.java]: This doesn't look right - you'll end up always setting the attributes to null.

Could just use `super.setAttributes(attributes != null ? Collections.synchronizedMap(attributes) : null);`

Comment (jdillon) [inline components/nexus-security/src/main/java/org/apache/shiro/nexus/NexusSessionFactory.java]: derpa derpa, you are right ;-)  maybe that is why this impl doesn't work at all ;-)

**Comment by jdillon** [comment]
fixed bug mentioned by @mcculls and auth seems to be *much* happier now.  Still cant' tell if this fixes the problem.  And also not seeing duplicate setAttributes() anymore which was likely due to the null bug.

**Comment by jdillon** [comment]
http://bamboo.s/browse/NX3-OSSF213-1 :white_check_mark:

**Comment by mcculls** [comment]
+1

**Comment by cstamas** [comment]
+1

**Comment by adreghiciu** [comment]
+1

**Comment by jdillon** [comment]
manually merged

**Pull request closed without merging**